### ISSUE TITLE
머딥 - 22d (CNN tensorflow)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#60]
# 목차
1. Conv2D Layer
2. Filter에 대한 이해
3. ReLU Layer
4. Maxpooling Layer
5. FC
6. 실습

데이터 준비
```
from tensorflow.keras import datasets
import numpy as np

mnist = datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

import matplotlib.pyplot as plt
import tensorflow as tf

image = X_train[0]
plt.imshow(image, 'gray')

image = image[tf.newaxis, ..., tf.newaxis]
image
```

## Conv2D Layer
1D 데이터 = stride가 세로 방향으로 움직인다 (텍스트) (char cnn)
2D 데이터 = stride가 두 방향(axis)으로 움직일 수 있다

- 매개변수
filters : 필터의 개수(c_out)
kernel_size : 필터의 크기(K_h, K_w)
strides : 몇 개의 픽셀을 건너뛸지
padding : 한줄 만든다(SAME), 안 만든다(VALID)
activation : 활성화함수 지정

### Conv2D를 쓰는 **2가지 방법**
```sage
tf.keras.layers.Conv2D(
    filters=3,
    kernel_size=(3, 3),
    strides=(1, 1),
    padding='SAME',
    activation='relu'    
    )

from tensorflow.keras.layers import Conv2D
Conv2D(
    filters=3,
    kernel_size=3,
    strides=1,
    padding='SAME',
    activation='relu'    
    )
```

이미지 준비
- 차원은 어떻게 될까?
```sage
image = tf.cast(image, dtype=tf.float32)
layer = tf.keras.layers.Conv2D(filters=5, kernel_size=3, strides=1, padding='SAME')
output = layer(image)
```
> padding을 넣었으므로 세로(output_H), 가로(output_W)는 안 바뀐다
> 출력 차원: N, output_H, output_W, C_out

- conv 시각화
```sage
fig, axes = plt.subplots(2,3, figsize=(10,10))
ax = axes.ravel()
ax[0].imshow(image[0,...,0], cmap='gray')
for i in range(5):
    ax[i+1].imshow(output[0,...,i], cmap='gray')
plt.show()
```

## filter? CNN이 이미지에서 **무엇을 보고** 있는가

- filter shape : (3, 3, 1, 5)
```sage
weights = layer.get_weights()
```

- weights[0] : 가중치가 들어 있음
- weights[1] : 편향이 들어 있음(=C_out 수)

```sage
print('Filter shape {}'.format(weights[0].shape))
print('Filter bias {}'.format(weights[1].shape))
```

- filter 시각화
```sage
_, axes = plt.subplots(1, 5, figsize=(10, 10))
for i, ax in enumerate(axes):
    ax.set_title('filter {}'.format(i + 1))
    ax.imshow(weights[0][..., 0, i], cmap='gray')
plt.show()
```

## ReLU Layer
```sage
np.min(output), np.max(output)

act_layer = tf.keras.layers.ReLU()
act_output = act_layer(output)

np.min(act_output), np.max(act_output)
```

- ReLU 시각화
```sage
_, axes = plt.subplots(1, 6, figsize=(10, 10))
axes[0].imshow(image[0, ..., 0], cmap='gray')
axes[0].set_title('original')

for idx, ax in enumerate(axes[1:]):
    ax.set_title('relu {}'.format(idx + 1))
    ax.imshow(act_output[0, ..., idx], cmap='gray')
plt.show()
```

## MaxPooling 
```sage
pool_layer = tf.keras.layers.MaxPool2D(pool_size=(2, 2), strides=(2, 2), padding='SAME')
pool_output = pool_layer(act_output)
```

- Pooling 시각화
```sage
_, axes = plt.subplots(1, 6, figsize=(10, 10))
axes[0].imshow(image[0, ..., 0], cmap='gray')
axes[0].set_title('original')

for idx, ax in enumerate(axes[1:]):
    ax.set_title('maxpooling {}'.format(idx + 1))
    ax.imshow(pool_output[0, ..., idx], cmap='gray')
plt.show()
```

## FC 와 결합(은닉층 + 출력층)
```sage
flatten_layer = tf.keras.layers.Flatten()
flatten_output = flatten_layer(pool_output)

# 은닉층
dense_layer = tf.keras.layers.Dense(32, activation='relu')
dense_output = dense_layer(flatten_output)

#출력층
dense_layer2 = tf.keras.layers.Dense(10, activation='softmax')
dense_output2 = dense_layer2(dense_output)
dense_output2
```

## 실습
- 조건
1 conv : 3, 3, 1, 32, padding='SAME'
relu
2 conv : 3, 3, 32, 32, padding='SAME'
relu
pooling : 2, 2
3 conv : 3, 3, 32, 64, padding='SAME'
relu
4 conv : 3, 3, 64, 64, padding='SAME'
relu
pooling : 2, 2
flatten
dense : 512, relu
dense : 10, softmax

```
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, ReLU, MaxPool2D, Flatten, Dense
```
maxpooling에 padding='SAME'을 넣으나 안 넣으나 parameter 개수에 변화가 없다??? 왜 ???
: padding이 작동하는 시점은 Conv Layer에 들어간 시점이기 때문

```sage
model = Sequential([
    Conv2D(input_shape=(28,28,1),filters=32, kernel_size=3, padding='SAME', activation='relu'),
    Conv2D(filters=32, kernel_size=3, padding='SAME'),
    ReLU(),
    MaxPool2D(pool_size=(2, 2)),
    Conv2D(filters=64, kernel_size=3, padding='SAME'),
    ReLU(),
    Conv2D(filters=64, kernel_size=3, padding='SAME'),
    ReLU(),
    MaxPool2D(pool_size=(2, 2)),
    Flatten(),
    Dense(512, activation='relu'),
    Dense(10, activation='softmax')
    ])
model.summary()
```

컴파일
```sage
model.compile(
    loss='sparse_categorical_crossentropy',
    optimizer='adam',
    metrics=['acc']
    )
```

- 채널 차원 추가 해줘야 한다!!!!
```sage
# ((60000, 28, 28), (60000,))
X_train.shape, y_train.shape

X_train = X_train[..., tf.newaxis]
X_test = X_test[..., tf.newaxis]
X_train = X_train/255.
X_test = X_test/255.
```

학습
```sage
model.fit(X_train,
          y_train,
          validation_data=(X_test, y_test),
          epochs=10)
```
